clients/src/test/java/org/apache/kafka/clients/consumer/internals/RequestFutureTest.java

```diff
@@ -14,14 +14,169 @@
 
 import org.junit.Test;
 
+import java.util.concurrent.atomic.AtomicBoolean;
+
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 
 public class RequestFutureTest {
 
     @Test
+    public void testBasicCompletion() {
+        RequestFuture<String> future = new RequestFuture<>();
+        String value = "foo";
+        future.complete(value);
+        assertTrue(future.isDone());
+        assertEquals(value, future.value());
+    }
+
+    @Test
+    public void testBasicFailure() {
+        RequestFuture<String> future = new RequestFuture<>();
+        RuntimeException exception = new RuntimeException();
+        future.raise(exception);
+        assertTrue(future.isDone());
+        assertEquals(exception, future.exception());
+    }
+
+    @Test
+    public void testVoidFuture() {
+        RequestFuture<Void> future = new RequestFuture<>();
+        future.complete(null);
+        assertTrue(future.isDone());
+        assertNull(future.value());
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void invokeCompleteAfterAlreadyComplete() {
+        RequestFuture<Void> future = new RequestFuture<>();
+        future.complete(null);
+        future.complete(null);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void invokeCompleteAfterAlreadyFailed() {
+        RequestFuture<Void> future = new RequestFuture<>();
+        future.raise(new RuntimeException());
+        future.complete(null);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void invokeRaiseAfterAlreadyFailed() {
+        RequestFuture<Void> future = new RequestFuture<>();
+        future.raise(new RuntimeException());
+        future.raise(new RuntimeException());
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void invokeRaiseAfterAlreadyCompleted() {
+        RequestFuture<Void> future = new RequestFuture<>();
+        future.complete(null);
+        future.raise(new RuntimeException());
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void invokeExceptionAfterSuccess() {
+        RequestFuture<Void> future = new RequestFuture<>();
+        future.complete(null);
+        future.exception();
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void invokeValueAfterFailure() {
+        RequestFuture<Void> future = new RequestFuture<>();
+        future.raise(new RuntimeException());
+        future.value();
+    }
+
+    @Test
+    public void listenerInvokedIfAddedBeforeFutureCompletion() {
+        RequestFuture<Void> future = new RequestFuture<>();
+
+        final AtomicBoolean wasInvoked = new AtomicBoolean(false);
```

Comment:
We probably want to have an `AtomicInteger` so that we check that the number of calls is right. Also, it probably makes sense to also track `onFailure` calls as there should be none. Similar for other tests like this.